File: test/sinceFilter.test.js
import test from "node:test";
import assert from "node:assert/strict";
import { getDonorNeighbourhoodWindowMs } from "../src/donorNeighbourhoodWindow.js";
import {
  filterRecordsSince,
  formatSinceQuery,
  getDonorListSinceMs,
  getInitiatorHistorySinceMs,
  parseSinceQuery,
  resolveListSinceMs
} from "../src/sinceFilter.js";

test("parseSinceQuery accepts hour minute day units", () => {
  assert.equal(parseSinceQuery("1h"), 3_600_000);
  assert.equal(parseSinceQuery("2h"), 7_200_000);
  assert.equal(parseSinceQuery("30m"), 1_800_000);
  assert.equal(parseSinceQuery(null), null);
  assert.equal(parseSinceQuery("bad"), null);
});

test("formatSinceQuery renders hour windows", () => {
  assert.equal(formatSinceQuery(getDonorListSinceMs()), "2h");
});

test("resolveListSinceMs caps donor to default window", () => {
  const windowMs = getDonorNeighbourhoodWindowMs();
  const historyMs = getInitiatorHistorySinceMs();
  assert.equal(resolveListSinceMs("donor", null), historyMs);
  assert.equal(resolveListSinceMs("donor", "7d"), historyMs);
  assert.equal(
    resolveListSinceMs("donor", "7d", { neighbourhoodMode: "near" }),
    windowMs
  );
  assert.equal(resolveListSinceMs("coordinator", null), null);
  assert.equal(resolveListSinceMs("coordinator", "2h"), 7_200_000);
});

test("filterRecordsSince keeps recent activity only", () => {
  const now = Date.parse("2026-06-02T12:00:00.000Z");
  const records = [
    { created_at: "2026-06-02T11:30:00.000Z" },
    { created_at: "2026-06-01T10:00:00.000Z" }
  ];
  const filtered = filterRecordsSince(records, 3_600_000, now);
  assert.equal(filtered.length, 1);
});

File: src/test/java/org/sharingbridge/integration/geo/SinceFilterTest.java
package org.sharingbridge.integration.geo;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNull;

import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Test;

class SinceFilterTest {

    @Test
    void parseSinceQueryAcceptsHourMinuteDayUnits() {
        assertEquals(3_600_000L, SinceFilter.parseSinceQuery("1h"));
        assertEquals(7_200_000L, SinceFilter.parseSinceQuery("2h"));
        assertEquals(1_800_000L, SinceFilter.parseSinceQuery("30m"));
        assertNull(SinceFilter.parseSinceQuery(null));
        assertNull(SinceFilter.parseSinceQuery("bad"));
    }

    @Test
    void formatSinceQueryRendersHourWindows() {
        assertEquals("2h", SinceFilter.formatSinceQuery(SinceFilter.getDonorListSinceMs()));
    }

    @Test
    void resolveListSinceMsCapsDonorToDefaultWindow() {
        long windowMs = SinceFilter.getDonorListSinceMs();
        long historyMs = SinceFilter.getInitiatorHistorySinceMs();
        assertEquals(historyMs, SinceFilter.resolveListSinceMs("donor", null, null));
        assertEquals(historyMs, SinceFilter.resolveListSinceMs("donor", "7d", null));
        assertEquals(windowMs, SinceFilter.resolveListSinceMs("donor", "7d", "near"));
        assertNull(SinceFilter.resolveListSinceMs("coordinator", null, null));
        assertEquals(7_200_000L, SinceFilter.resolveListSinceMs("coordinator", "2h", null));
    }

    @Test
    void filterRecordsSinceKeepsRecentActivityOnly() {
        long now = java.time.Instant.parse("2026-06-02T12:00:00.000Z").toEpochMilli();
        List<Map<String, Object>> records =
                List.of(
                        Map.of("created_at", "2026-06-02T11:30:00.000Z"),
                        Map.of("created_at", "2026-06-01T10:00:00.000Z"));
        List<Map<String, Object>> filtered = SinceFilter.filterRecordsSince(records, 3_600_000L, now);
        assertEquals(1, filtered.size());
    }
}
